### PR TITLE
Add ucrt64 mingw toolchain.

### PR DESCRIPTION
**Pull request by angerman**
this is mingw/gcc/ucrt the llvm based one is called clang64 as per https://www.msys2.org/docs/environments/.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by angerman** [comment]
```
$ strings $(nix build -f . pkgs.pkgsCross.ucrt64.hello --no-
link --print-out-paths)/bin/hello.exe | grep '\.dll'
Winsock.dll version out of range
KERNEL32.dll
api-ms-win-crt-convert-l1-1-0.dll
api-ms-win-crt-environment-l1-1-0.dll
api-ms-win-crt-filesystem-l1-1-0.dll
api-ms-win-crt-heap-l1-1-0.dll
api-ms-win-crt-locale-l1-1-0.dll
api-ms-win-crt-math-l1-1-0.dll
api-ms-win-crt-private-l1-1-0.dll
api-ms-win-crt-runtime-l1-1-0.dll
api-ms-win-crt-stdio-l1-1-0.dll
api-ms-win-crt-string-l1-1-0.dll
api-ms-win-crt-time-l1-1-0.dll
```
vs
```
$ strings $(nix build -f . pkgs.pkgsCross.mingwW64.hello --n
o-link --print-out-paths)/bin/hello.exe | grep '\.dll'
Winsock.dll version out of range
KERNEL32.dll
msvcrt.dll
```

**Comment by angerman** [comment]
@shlevy did open #171418 a while ago, but eventually got reverted in #173487; This PR explicitly tries to _only_ do the UCRT part with GCC for now.

**Comment by Ericson2314** [comment]
Can you make the first commit switch all those conditionals to `isMinGW` so we don't need to enumerate both libcs, and then the second commit actually add ucrt?

**Comment by angerman** [comment]
> Can you make the first commit switch all those conditionals to `isMinGW` so we don't need to enumerate both libcs, and then the second commit actually add ucrt?

Done.

**Comment by angerman** [comment]
Why can't I assign this to myself?

**Comment by nagy** [comment]
Can you also please follow the commit message conventions at https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions?

**Comment by angerman** [comment]
> Can you also please follow the commit message conventions at https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions?

@nagy Let me know if this is better. Feel free to suggest even better ones!

**Comment by nagy** [comment]
those look ok

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -20687,7 +20687,7 @@ with pkgs;
   # These are used when buiding compiler-rt / libgcc, prior to building libc.
   preLibcCrossHeaders = let
     inherit (stdenv.targetPlatform) libc;
-  in     if libc == "msvcrt" then targetPackages.windows.mingw_w64_headers or windows.mingw_w64_headers
+  in     if isMinGW then targetPackages.windows.mingw_w64_headers or windows.mingw_w64_headers
```
Comment (OPNA2608) [inline pkgs/top-level/all-packages.nix:20690]:
```
error: undefined variable 'isMinGW'
```

```suggestion
  in     if stdenv.targetPlatform.isMinGW then targetPackages.windows.mingw_w64_headers or windows.mingw_w64_headers
```

Comment (angerman) [reply pkgs/top-level/all-packages.nix:20690]: @OPNA2608 thanks, missed that in the rewrite.

**Comment by Artturin** [comment]
Should target master instead of a release branch.

File: pkgs/development/compilers/gcc/10/default.nix
Comment (Artturin) [inline pkgs/development/compilers/gcc/10/default.nix:1]: The `gcc/*/default.nix` files don't exist anymore, `crossMingw` variable is now in `pkgs/development/compilers/gcc/common/configure-flags.nix`

Comment (angerman) [reply pkgs/development/compilers/gcc/10/default.nix:1]: Fun, this is going to be fun back porting then.

Comment (angerman) [reply pkgs/development/compilers/gcc/10/default.nix:1]: @Artturin rebased against master now.

**Pull request merged**